vividus-plugin-browserstack/src/main/java/org/vividus/selenium/browserstack/BrowserStackLocalManager.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2019-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.selenium.browserstack;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.browserstack.local.Local;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.vividus.selenium.tunnel.TunnelManager;
+import org.vividus.testcontext.TestContext;
+import org.vividus.util.ResourceUtils;
+import org.vividus.util.wait.RetryTimesBasedWaiter;
+
+public class BrowserStackLocalManager implements TunnelManager<BrowserStackLocalOptions>
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(BrowserStackLocalManager.class);
+    private static final Object KEY = BrowserStackLocalConnection.class;
+
+    private final String sauceLabsAccessKey;
+    private final TestContext testContext;
+
+    public BrowserStackLocalManager(String sauceLabsAccessKey, TestContext testContext)
+    {
+        this.sauceLabsAccessKey = sauceLabsAccessKey;
+        this.testContext = testContext;
+    }
+
+    private final Map<BrowserStackLocalOptions, BrowserStackLocalConnection> activeConnections = new HashMap<>();
+
+    @Override
+    public String start(BrowserStackLocalOptions options)
+    {
+        BrowserStackLocalConnection connection = getCurrentConnection();
+        if (connection == null)
+        {
+            try
+            {
+                connection = activeConnections.get(options);
+                if (connection == null)
+                {
+                    synchronized (activeConnections)
+                    {
+                        connection = activeConnections.get(options);
+                        if (connection != null)
+                        {
+                            putCurrentConnection(connection);
+                            connection.waitForStart();
+                            return connection.getLocalIdentifier();
+                        }
+                        connection = new BrowserStackLocalConnection(options);
+                        putCurrentConnection(connection);
+                        activeConnections.put(options, connection);
+                    }
+                }
+                connection.startConnection();
```

Comment:
why is it outside `synchronized` block?

Reply:
each thread with different options has its own Local instance, if options are the same, other threads will not reach this point

Reply:
```java
connection = activeConnections.get(options);
if (connection == null)
{
    synchronized (activeConnections)
    {
        connection = activeConnections.get(options);
        if (connection == null)
        {
            connection = new BrowserStackLocalConnection(options);
            connection.startConnection();
            connection.waitForStart();
            activeConnections.put(options, connection);
        }
        putCurrentConnection(connection);
    }
}
```